src/messaging/Channel.h

```diff
@@ -0,0 +1,248 @@
+        kNetwork_LowLatency,
+        kNetwork_LowPowerConsumption,
+    };
+
+    enum class TransportPreference
+    {
+        kTransport_Connectionless,
+        kTransport_PreferConnectionOriented, // will fallback to connectionless TCP is not supported
+        kTransport_ConnectionOriented,       // will fail if TCP is not supported
+
+        kTransport_Default = kTransport_Connectionless,
+    };
+
+    enum class SessionType
+    {
+        kSession_PASE, // Use SPAKE2 key exchange
+        kSession_CASE, // Use SIGMA key exchange
+
+        kSession_Default = kSession_CASE,
+    };
+
+    ChannelBuilder & SetPeerNodeId(NodeId peerNodeId)
+    {
+        mPeerNodeId = peerNodeId;
+        return *this;
+    }
+    NodeId GetPeerNodeId() const { return mPeerNodeId; }
+
+    ChannelBuilder & SetNetworkPreference(NetworkPreference preference)
+    {
+        mNetworkPreference = preference;
+        return *this;
+    }
+    NetworkPreference GetNetworkPreference() const { return mNetworkPreference; }
+
+    ChannelBuilder & SetTransportPreference(TransportPreference preference)
+    {
+        mTransportPreference = preference;
+        return *this;
+    }
+    TransportPreference GetTransportPreference() const { return mTransportPreference; }
+
+    ChannelBuilder & SetSessionType(SessionType preference)
+    {
+        mSessionType = preference;
+        return *this;
+    }
+    SessionType GetSessionType() const { return mSessionType; }
+
+    uint16_t GetPeerKeyID() const { return mCaseParameters.mPeerKeyId; }
+    ChannelBuilder & SetPeerKeyID(uint16_t keyId)
+    {
+        if (mSessionType == SessionType::kSession_CASE)
+            mCaseParameters.mPeerKeyId = keyId;
+        return *this;
+    }
+
+    uint32_t GetPeerSetUpPINCode() const { return mPaseParameters.mPeerSetUpPINCode; }
+    ChannelBuilder & SetPeerSetUpPINCode(uint32_t peerPINCode)
+    {
+        if (mSessionType == SessionType::kSession_PASE)
+            mPaseParameters.mPeerSetUpPINCode = peerPINCode;
+        return *this;
+    }
+
+#ifndef NDEBUG
+    Optional<Inet::IPAddress> GetHintPeerAddress() const { return mHintPeerAddr; }
+    ChannelBuilder & SetHintPeerAddress(Inet::IPAddress peerAddr)
+    {
+        mHintPeerAddr.SetValue(peerAddr);
+        return *this;
+    }
+#endif // NDEBUG
+
+private:
+    NodeId mPeerNodeId                       = kUndefinedNodeId;
+    NetworkPreference mNetworkPreference     = NetworkPreference::kNetwork_Default;
+    TransportPreference mTransportPreference = TransportPreference::kTransport_Default;
+    SessionType mSessionType                 = SessionType::kSession_Default;
+    union
+    {
+        struct
+        {
+            uint16_t mPeerKeyId;
+        } mCaseParameters;
+        struct
+        {
+            uint32_t mPeerSetUpPINCode;
+        } mPaseParameters;
+    };
+
+#ifndef NDEBUG
+    Optional<Inet::IPAddress> mHintPeerAddr;
+#endif // NDEBUG
+};
+
+class ExchangeContext;
```

Comment:
could we #include instead of forward declare?

forward declaration can hide layering issues that the buildsystem would catch (build system would complain about circular dependencies in includes, but cannot detect this if using forward declarations)

Reply:
This is indeed a circular dependences, in `ChannelHandle` class we have:

```
ExchangeContext * ChannelHandle::NewExchange(ExchangeDelegate * delegate)
```

In `ExchangeContext` class, we will have (not implemented in this PR yes):

```
ChannelHandle ExchangeContext::GetChannel()
```

The forward declaration is required to break the loop. IMHO, it is fine to have circular dependencies like this one in the same module. but different module can't depend each other.

Reply:
I would call it out in a comment then.

It may be unavoidable in this case, however generally every circular dependency either shows a design issoe or a very tight coupling (we have tight coupling here).